### PR TITLE
machines: Fix issue with Virtual Network being configured with a reserved address

### LINKED ISSUE #13506: Network: Wrong description in the "Create a network section" can lead into strange network issues
**Basic information**

Name         : cockpit
Version      : 210
OS/Release      : 1.fc30
Architecture : x86_64
Size         : 53 k
Quelle       : cockpit-210-1.fc30.src.rpm
Page: Host --> Virtual Machines --> Networks --> Create Virtual Network

**Description**

While creating a new virtual network, it automatically fills the "IPv4 Network" field with 192.168.100.**0**, which is ok but after creating the network, it takes the address 192.168.100.0 as a gateway address in the virtual machines connected to the network and IP for the virtualization host.

In my case the basic connectivity was given but I had very strange network issues between docker container and other VM's connecting to it until I changed it manually with virsh net-edit to 192.168.100.1. 

IMO it would be better to change the field to gateway and use the first available IP in the given range. 

**Steps to reproduce**

1. Go to Host --> Virtual Machines --> Networks --> Create Virtual Network
2. Add the network with the Address autmatically filled

cockpit-network

ipnewnet

### PR DESCRIPTION
**Pull request by KKoukiou**
When creating a Virtual Network make sure to use as 'address' the first
available host address calculated from the network IP and the netmask.

Use ipaddress python lib so that we can verify that the address got
correctly calculated in the tests.

Fixes #13506

### CODE REVIEW

File: pkg/machines/components/networks/createNetworkDialog.jsx [hunk 1]
```diff
@@ -403,10 +403,16 @@ class CreateNetworkModal extends React.Component {
             const ipv6 = ip === "IPv4 only" ? undefined : this.state.ipv6;
             const ipv4 = ip === "IPv6 only" ? undefined : this.state.ipv4;
             const netmask = utils.netmaskConvert(this.state.netmask);
+            // If you take an IP address an apply the network mask (bitwise and) then you get the network address - no host can have such address
+            // Increment by 1 to get the first usable address of the block
+            const ipv4FirstUsableAddress = (ipv4 &&
+                ipv4.split(".")
+                        .map((item, index) => (item & netmask.split(".")[index]) + Number(index == 3))
+                        .join("."));
```
Comment (skobyda) [inline pkg/machines/components/networks/createNetworkDialog.jsx]: Could this helper function maybe be moved into `pkg/machines/components/networks/utils.js`?

File: pkg/machines/components/networks/createNetworkDialog.jsx [hunk 2]
```diff
@@ -403,10 +403,16 @@ class CreateNetworkModal extends React.Component {
...
 
             this.setState({ createInProgress: true });
             networkCreate({
-                connectionName, name, forwardMode, device, ipv4, netmask, ipv6, prefix,
+                connectionName, name, forwardMode, device, ipv4: ipv4FirstUsableAddress, netmask, ipv6, prefix,
```
Comment (skobyda) [inline pkg/machines/components/networks/createNetworkDialog.jsx]: Let me understand this properly, what this code does is that it simply takes ANY ip address and user inputted in IPv4 Network input field and increases it by one?

Shouldn't we change the default address in the UI to "192.168.100.1" so what user sees in UI corresponds with what's then put into XML?

Also what would happen if user for whatever reason put address which ends with .255. Would it just increase by one and jump out of valid IPv6 range?

Comment (KKoukiou) [reply pkg/machines/components/networks/createNetworkDialog.jsx]: That's what virt-manager does - just increases by one I think. (playing around with the UI at least gave me this impression)
Good point for the address which ends with .255. We should not allow that I guess.

Comment (skobyda) [reply pkg/machines/components/networks/createNetworkDialog.jsx]: To me it seems that virt-manager doesn't increase it by one. To me it seems it ignores whatever you put into last part of IPv4 address and just replaces it with .1
Example: in virt-manager, I tried to create virtual network with 192.168.100.77/24, and virt-manager put 192.168.100.1 into XML.
So to me it seems it just takes the part defined by submask from user input field (the first 24 bits in this case) and overwrites rest with 1

File: pkg/machines/components/networks/createNetworkDialog.jsx (hunk 1 above)
Comment (KKoukiou) [reply pkg/machines/components/networks/createNetworkDialog.jsx]: I don't reuse it anywhere but I can move it, sure.

File: pkg/machines/components/networks/createNetworkDialog.jsx [hunk 3]
```diff
@@ -50,6 +50,8 @@ function validateParams(dialogValues) {
             validationFailed.netmask = _("Mask or Prefix Length should not be empty");
         else if (!utils.validateNetmask(dialogValues.netmask))
             validationFailed.netmask = _("Invalid IPv4 mask or prefix length");
+        else if (!utils.ipv4FirstUsableAddress(dialogValues.ipv4, dialogValues.netmask))
+            validationFailed.netmask = _("There are not any usable addresses within this subnet");
```
Comment (martinpitt) [inline pkg/machines/components/networks/createNetworkDialog.jsx]: "There are no usable addresses.."

File: pkg/machines/components/networks/utils.js [hunk 4]
```diff
@@ -229,3 +229,34 @@ export function isIpv6InNetwork(network, prefix, ip) {
...
+ * @param {string} netmask
+ */
+export function ipv4FirstUsableAddress(ipv4, netmask) {
+    // Within a subnet, two host addresses - all-zeros and one all-ones are reserved as network address and broadcast, respectively.
+    if (!ipv4)
+        return undefined;
+
+    let mask = netmask;
+    if (netmask.split(".").length == 4)
+        mask = netmaskToCIDR(netmask);
```
Comment (martinpitt) [inline pkg/machines/components/networks/utils.js]: Can it happen that ipv4 is in an invalid format here? like, less or more than 4 components, numbers <1 or > 254, or not numbers at all, etc? I. e. does this function need to do that structural validation, or is that already done elsewhere?

File: pkg/machines/components/networks/utils.js [hunk 5]
```diff
@@ -229,3 +229,34 @@ export function isIpv6InNetwork(network, prefix, ip) {
...
+    // For a /31 subnet the number of usable addresses is zero.
```
Comment (martinpitt) [inline pkg/machines/components/networks/utils.js]: And likewise, if it's 0 then it's invalid as well.

File: pkg/machines/components/networks/utils.js [hunk 6]
```diff
@@ -229,3 +229,34 @@ export function isIpv6InNetwork(network, prefix, ip) {
...
+    if (mask > 30)
+        return undefined;
+
+    return ipv4.split(".")
+            .map((item, index) => (item & netmask.split(".")[index]) + Number(index == 3))
```
Comment (martinpitt) [inline pkg/machines/components/networks/utils.js]: This looks overcomplicated and error prone. At this point you should already have made sure that it's not an invalid IP address, either by existing earlier validation, or by doing additional validation yourself. It's enough to split at the last dot and increase that, or keep doing what you do, and just change the `[3]` element. Also, you need to handle the case that the last component is 254, as 255 is not a valid IP (it's the broadcast).

What is that IP exactly, can it really be arbitrary? In the issue the screenshot shows ".0", which isn't an IP, but just gives a network. Will it always be 0? Can it be some actual IP? Do the two cases really mean the same?

File: pkg/machines/components/networks/utils.js [hunk 7]
```diff
@@ -229,3 +229,34 @@ export function isIpv6InNetwork(network, prefix, ip) {
...
+            .join(".");
+}
+
+function netmaskToCIDR(mask) {
+    const maskNodes = mask.match(/(\d+)/g);
```
Comment (martinpitt) [inline pkg/machines/components/networks/utils.js]: `.split('.')` ?

File: pkg/machines/components/networks/utils.js [hunk 8]
```diff
@@ -229,3 +229,34 @@ export function isIpv6InNetwork(network, prefix, ip) {
...
+    let cidr = 0;
+    for (const i in maskNodes)
+        cidr += (((maskNodes[i] >>> 0).toString(2)).match(/1/g) || []).length;
```
Comment (martinpitt) [inline pkg/machines/components/networks/utils.js]: `s >>> 0` is just a weird way to write `Number(s)`. This takes a while to realize what it does, and it's concise (albeit probably the most inefficient way to count bits, but performance doesn't matter much). Please do add a comment about "count 1 bits in the mask".

But again there is not enough error checking here. What if the netmask isn't a string of 1's followed by all zeros? E. g. if someone specifies a netmask like "255.255.129.0", this function would happily turn it into a CIDR mask of 18, although it's not a valid mask. Or what happens on "even wronger" input?

**Pull request closed without merging**